btclightclient/partial_merkle_tree.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// / We verify partial merkle tree proof. This logic are using in verifytxoutproof
+// / We are follow the logic in verifytxoutproof.
+
+type PartialMerkleTree struct {
+	numberTransactions uint
+	vBits              []bool
+	vHash              []*chainhash.Hash
+}
+
+type MerkleProof struct {
+	nodeValue  chainhash.Hash
+	merklePath []chainhash.Hash
+	pos        uint32
+}
+
+// TODO: verify this value
+const maxAllowBytes = 65536
+
+// TODO: wrap error
+func readPartialMerkleTree(r io.Reader, buf []byte) (*PartialMerkleTree, error) {
+	if _, err := io.ReadFull(r, buf[:4]); err != nil {
+		return nil, err
+	}
+
+	numberTransactions := uint(binary.LittleEndian.Uint32(buf[:4]))
+
+	var pver uint32 // pversion but btcd don't use this in those function we want.
+
+	var vHash []*chainhash.Hash
+	if numberHash, err := wire.ReadVarIntBuf(r, pver, buf); err != nil {
+		return nil, err
+	} else {
+		if numberHash*32 > uint64(maxAllowBytes) {
+			return nil, errors.New("number of hash is too big")
+		}
+
+		bytes := make([]byte, numberHash*32)
+		_, err := io.ReadFull(r, bytes)
+
+		if err != nil {
+			return nil, err
+		}
+
+		vHash = make([]*chainhash.Hash, numberHash)
+		for i := 0; i < int(numberHash); i++ {
+			hash, err := chainhash.NewHash(bytes[i*32 : (i+1)*32])
+			if err != nil {
+				return nil, err
+			}
+			vHash[i] = hash
+		}
+	}
+
+	var vBits []bool
+	if vBytes, err := wire.ReadVarBytes(r, pver, uint32(maxAllowBytes), "vBits"); err != nil {
+		return nil, err
+	} else {
+
+		vBits = make([]bool, len(vBytes)*8)
+		i := 0
+		for _, b := range vBytes {
+			for j := 0; j < 8; j++ {
+				vBits[i] = (b & (1 << j)) != 0
+				i++
+			}
+		}
+	}
+
+	parialMerkleTree := PartialMerkleTree{
+		numberTransactions: numberTransactions,
+		vBits:              vBits,
+		vHash:              vHash,
+	}
+
+	return &parialMerkleTree, nil
+}
+
+func ParialMerkleTreeFromBytes(hexStr string) (*PartialMerkleTree, error) {
+	hexBytes, err := hex.DecodeString(hexStr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	reader := bytes.NewReader(hexBytes)
+
+	pmk, err := readPartialMerkleTree(reader, hexBytes)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return pmk, nil
+}
+
+func (pmk *PartialMerkleTree) CalcTreeWidth(height uint32) uint {
+	return (pmk.numberTransactions + (1 << height) - 1) >> height
+}
+
+func (pmk *PartialMerkleTree) Height() uint32 {
+	var nHeight uint32 = 0
+	for pmk.CalcTreeWidth(nHeight) > 1 {
+		nHeight++
+	}
+	return nHeight
+}
+
+/// Port logic from gettxoutproof from bitcoin-core
+/// TODO: Make error handler more sense
+func (pmk *PartialMerkleTree) computeMerkleProofRecursive(height, pos uint32, nBitUsed, nHashUsed *uint32, txID *chainhash.Hash) (*MerkleProof, error) {
+	if int(*nBitUsed) >= len(pmk.vBits) {
+		return nil, errors.New("Error")
+	}
+
+	fParentOfMatch := pmk.vBits[*nBitUsed]
+	*nBitUsed = *nBitUsed + 1
+
+	if height == 0 || !fParentOfMatch {
```

Comment:
if height is 0, then shouldn't we return?

Reply:
This is or so if `height` is zero we still jump into this scope.